### PR TITLE
[SYCL][Doc] Update Sphinx and dependencies

### PR DESCRIPTION
**Pull request by alexbatashev**
recommonmark (which was used for mardown rendering) is deprecated and
will no longer be actively maintained. MyST and Sphinx 4 offer better
Markdown and C++ syntax highlighting support.

### CODE REVIEW

File: sycl/doc/PreprocessorMacros.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 This file describes macros that have effect on SYCL compiler and run-time.
 
-### `RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR`
+## `RESTRICT_WRITE_ACCESS_TO_CONSTANT_PTR`
```
Comment (alexbatashev) [inline sycl/doc/PreprocessorMacros.md:5]: MyST warns when header order is violated, so I had to update this document

Comment (bader) [reply sycl/doc/PreprocessorMacros.md:5]: Probably using headers here is not the best idea. Can we switch to the list and make macro names bold?

Comment (alexbatashev) [reply sycl/doc/PreprocessorMacros.md]: Does it look better now?

Comment (bader) [reply sycl/doc/PreprocessorMacros.md]: Yes. Thank you.

**Review by bader — APPROVED** [review]

**Pull request merged**